plugins/auth/src/pam.cpp

```diff
@@ -0,0 +1,373 @@
+                THROW(SYS_INVALID_INPUT_PARAM,
+                      fmt::format("[{}:{}] missing [{}] in request", __func__, __LINE__, n));
+            }
+
+            rodsServerHost_t* host = nullptr;
+
+            log_auth::debug("connecting to catalog provider");
+
+            if (const int ec = getAndConnRcatHost(&comm, MASTER_RCAT, comm.clientUser.rodsZone, &host); ec < 0) {
+                THROW(ec, "getAndConnRcatHost failed.");
+            }
+
+            if (LOCAL_HOST != host->localFlag) {
+                const auto disconnect = irods::at_scope_exit{[host]
+                    {
+                        rcDisconnect(host->conn);
+                        host->conn = nullptr;
+                    }
+                };
+
+                log_auth::debug("redirecting call to CSP");
+
+                if (const auto ec = sslStart(host->conn); ec) {
+                    THROW(ec, "could not establish SSL connection");
+                }
+
+                const auto end_ssl = irods::at_scope_exit{[host] { sslEnd(host->conn); }};
+
+                return request(*host->conn, req);
+            }
+
+            json resp{req};
+
+            log_auth::debug("getting TTL param");
+
+            int ttl = 0;
+            if (req.contains(irods::AUTH_TTL_KEY)) {
+                if (const auto& ttl_str = req.at(irods::AUTH_TTL_KEY).get_ref<const std::string&>();
+                    !ttl_str.empty()) {
+                    try {
+                        ttl = boost::lexical_cast<int>(ttl_str);
+                    }
+                    catch (const boost::bad_lexical_cast& e) {
+                        THROW(SYS_INVALID_INPUT_PARAM, fmt::format("invalid TTL [{}]", ttl_str));
+                    }
+                }
+            }
+
+            const auto username = req.at("user_name").get_ref<const std::string&>();
+            const auto password = req.at(irods::AUTH_PASSWORD_KEY).get_ref<const std::string&>();
+
+            log_auth::trace("performing PAM auth check for [{}]", username);
+
+            run_pam_auth_check(username, password);
+
+            log_auth::trace("PAM auth check done", username);
+
+            char password_out[MAX_NAME_LEN]{};
+            char* pw_ptr = &password_out[0];
+            const int ec = chlUpdateIrodsPamPassword(&comm,
+                                                     const_cast<char*>(username.c_str()),
+                                                     ttl,
+                                                     nullptr,
+                                                     &pw_ptr);
+            if (ec < 0) {
+                THROW(ec, "failed updating iRODS pam password");
+            }
+
+            resp["request_result"] = password_out;
+
+            if (comm.auth_scheme) {
+                free(comm.auth_scheme);
+            }
+
+            static constexpr char* auth_scheme_pam = "pam";
+            comm.auth_scheme = strdup(auth_scheme_pam);
```

Comment:
Most of this was lifted from the legacy plugin and adapted to the new framework.